File: src/web/magmac/app/lang/node/MultipleCaseValue.ts
import { Option } from "../../../../magmac/api/Option";
import { List } from "../../../../magmac/api/collect/list/List";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { NodeListRule } from "../../../../magmac/app/compile/rule/NodeListRule";
import { PrefixRule } from "../../../../magmac/app/compile/rule/PrefixRule";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { StripRule } from "../../../../magmac/app/compile/rule/StripRule";
import { SuffixRule } from "../../../../magmac/app/compile/rule/SuffixRule";
import { TypeRule } from "../../../../magmac/app/compile/rule/TypeRule";
import { StatementFolder } from "../../../../magmac/app/compile/rule/fold/StatementFolder";
import { Destructors } from "../../../../magmac/app/lang/Destructors";
import { JavaFunctionSegment } from "../../../../magmac/app/lang/java/JavaFunctionSegment";
export class MultipleCaseValue {
	public static deserialize( node : Node) : Option<CompileResult<CaseValue>> {return Destructors.destructWithType( "case-multiple", node).map( 0);;}
	public static createRule( segment : Rule) : TypeRule {return new TypeRule( "case-multiple", new StripRule( new PrefixRule( "{", new SuffixRule( NodeListRule.createNodeListRule( "children", new StatementFolder( ), segment), "}"))));;}
}

File: packages/compiler-java/src/main/java/magmac/app/lang/node/MultipleCaseValue.java
package magmac.app.lang.node;

/**
 * Represents a case body containing multiple statements.
 * Example syntax:
 * {@code case 1 {
 *     doSomething();
 * }}
 */

import magmac.api.Option;
import magmac.api.collect.list.List;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.NodeListRule;
import magmac.app.compile.rule.PrefixRule;
import magmac.app.compile.rule.Rule;
import magmac.app.compile.rule.StripRule;
import magmac.app.compile.rule.SuffixRule;
import magmac.app.compile.rule.TypeRule;
import magmac.app.compile.rule.fold.StatementFolder;
import magmac.app.lang.Destructors;
import magmac.app.lang.java.JavaFunctionSegment;

public record MultipleCaseValue(List<JavaFunctionSegment> children) implements CaseValue {
    public static Option<CompileResult<CaseValue>> deserialize(Node node) {
        return Destructors.destructWithType("case-multiple", node).map(destructor -> destructor.withNodeList("children", FunctionSegments::deserialize)
                .complete(MultipleCaseValue::new));
    }

    public static TypeRule createRule(Rule segment) {
        return new TypeRule("case-multiple", new StripRule(new PrefixRule("{", new SuffixRule(NodeListRule.createNodeListRule("children", new StatementFolder(), segment), "}"))));
    }
}
